modules/Sources/Dependencies/AddressBookClient/AddressBookEncryption.swift

```diff
@@ -0,0 +1,286 @@
+    static func encryptContacts(_ abContacts: AddressBookContacts) throws -> Data {
+        @Dependency(\.walletStorage) var walletStorage
+        
+        guard let encryptionKeys = try? walletStorage.exportAddressBookEncryptionKeys(), let addressBookKey = encryptionKeys.getCached(account: 0) else {
+            throw AddressBookClient.AddressBookClientError.missingEncryptionKey
+        }
+        
+        var encryptionVersionData = Data()
+        encryptionVersionData.append(contentsOf: intToBytes(AddressBookEncryptionKeys.Constants.version))
+        
+        var dataForEncryption = Data()
+        
+        // Serialize `address book version`
+        dataForEncryption.append(contentsOf: intToBytes(abContacts.version))
+        
+        // Serialize `lastUpdated`
+        dataForEncryption.append(contentsOf: AddressBookClient.serializeDate(Date()))
+        
+        // Serialize `contacts.count`
+        dataForEncryption.append(contentsOf: intToBytes(abContacts.contacts.count))
+        
+        // Serialize `contacts`
+        abContacts.contacts.forEach { contact in
+            let serializedContact = serializeContact(contact)
+            dataForEncryption.append(serializedContact)
+        }
+        
+        // Generate a fresh one-time sub-key for encrypting the address book.
+        let salt = SymmetricKey(size: SymmetricKeySize.bits256)
+        
+        return try salt.withUnsafeBytes { salt in
+            let salt = Data(salt)
+            let subKey = addressBookKey.deriveEncryptionKey(salt: salt)
+            
+            // Encrypt the serialized address book.
+            // CryptoKit encodes the SealedBox as `nonce || ciphertext || tag`.
+            let sealed = try ChaChaPoly.seal(dataForEncryption, using: subKey)
+            
+            // Prepend the encryption version & salt to the SealedBox so we can re-derive the sub-key.
+            
+            // unencrypted data
+            return encryptionVersionData + salt
+            // encrypted data
+            + sealed.combined
+        }
+    }
+    
+    /// Tries to decrypt the data with the structure:
+    ///     [Unencrypted data]    `encryption version`
+    ///     [Unencrypted data]    `salt`
+    ///     [Encrypted data]        `address book version`
+    ///     [Encrypted data]        `timestamp`
+    ///     [Encrypted data]        `contacts`
+    static func contactsFrom(encryptedData: Data) throws -> AddressBookContacts {
+        @Dependency(\.walletStorage) var walletStorage
+        
+        guard let encryptionKeys = try? walletStorage.exportAddressBookEncryptionKeys(), let addressBookKey = encryptionKeys.getCached(account: 0) else {
+            throw AddressBookClient.AddressBookClientError.missingEncryptionKey
+        }
+        
+        var offset = 0
+        
+        // Deserialize `encryption version`
+        let encryptionVersionBytes = encryptedData.subdata(in: offset..<(offset + MemoryLayout<Int>.size))
```

Comment:
`subdata(in:)` appears in practice to be memory-unsafe (https://www.programiz.com/online-compiler/0eYmDmico8t92). So for example, if the length of `encryptedData` were less than 4 bytes, this would be unsafe. What I would do in Rust is write a safe helper method that wraps the unsafe one, and only use the helper.

Reply:
Resolved with custom

```swift
    private static func subdata(of data: Data, in range: Range<Data.Index>) throws -> Data {
        guard data.count >= range.lowerBound.distance(to: range.upperBound) else {
            throw AddressBookClient.AddressBookClientError.subdataRange
        }
        
        return data.subdata(in: range)
    }
```